libevmcore/Assembly.cpp

```diff
@@ -156,6 +163,137 @@ ostream& Assembly::stream(ostream& _out, string const& _prefix, StringMap const&
+Json::Value Assembly::createJsonValue(string _name, int _begin, int _end, string _value, string _jumpType) const
+{
+	Json::Value value;
+	value["name"] = _name;
+	value["begin"] = _begin;
+	value["end"] = _end;
+	if (!_value.empty())
+		value["value"] = _value;
+	if (!_jumpType.empty())
+		value["jumpType"] = _jumpType;
+	return value;
+}
+
+Json::Value Assembly::streamAsmJson(ostream& _out, string const& _prefix, StringMap const& _sourceCodes, bool _inJsonFormat) const
+{
+	Json::Value root;
+
+	string currentArrayName = ".code";
+	std::vector<Json::Value> currentCollection;
+	for (AssemblyItem const& i: m_items)
+	{
+		switch (i.type())
+		{
+		case Operation:
+			currentCollection.push_back(
+				createJsonValue(instructionInfo(i.instruction()).name, i.getLocation().start, i.getLocation().end, i.getJumpTypeAsString()));
+			break;
+		case Push:
+		{
+			std::stringstream hexStr;
+			hexStr << hex << i.data();
+			currentCollection.push_back(
+				createJsonValue(string("PUSH"), i.getLocation().start, i.getLocation().end, hexStr.str(), i.getJumpTypeAsString()));
+		}
+			break;
+		case PushString:
+			currentCollection.push_back(
+				createJsonValue(string("PUSH tag"), i.getLocation().start, i.getLocation().end, m_strings.at((h256)i.data())));
+			break;
+		case PushTag:
+		{
+			std::stringstream hexStr;
+			hexStr << hex << i.data();
+			currentCollection.push_back(
+				createJsonValue(string("PUSH [tag]"), i.getLocation().start, i.getLocation().end, hexStr.str()));
+		}
+			break;
+		case PushSub:
+			currentCollection.push_back(
+				createJsonValue(string("PUSH [$]"), i.getLocation().start, i.getLocation().end, dev::toString(h256(i.data()))));
+			break;
+		case PushSubSize:
+			currentCollection.push_back(
+				createJsonValue(string("PUSH #[$]"), i.getLocation().start, i.getLocation().end, dev::toString(h256(i.data()))));
+			break;
+		case PushProgramSize:
+			currentCollection.push_back(
+				createJsonValue(string("PUSHSIZE"), i.getLocation().start, i.getLocation().end));
+			break;
+		case Tag:
+		{
+			Json::Value collection(Json::arrayValue);
+			for (auto it: currentCollection)
+				collection.append(it);
+			currentCollection.clear();
+			root[currentArrayName] = collection;
+			currentArrayName = "tag" + string(i.data());
+			Json::Value jumpdest;
+			jumpdest["name"] = "JUMDEST";
+			currentCollection.push_back(jumpdest);
+		}
+			break;
+		case PushData:
+		{
+			Json::Value pushData;
+			pushData["name"] = "PUSH hex";
+			std::stringstream hexStr;
+			hexStr << hex << i.data();
+			currentCollection.push_back(createJsonValue(string("PUSH hex"), i.getLocation().start, i.getLocation().end, hexStr.str()));
+		}
+			break;
+		default:
+			BOOST_THROW_EXCEPTION(InvalidOpcode());
+		}
+	}
+
+	Json::Value collection(Json::arrayValue);
+	for (auto it: currentCollection)
+		collection.append(it);
+	root[currentArrayName] = collection;
+
+	if (!m_data.empty() || !m_subs.empty())
+	{
+		Json::Value dataCollection(Json::arrayValue);
```

Comment:
I think `dataCollection` should be a hashtable whose keys are the keys of m_data (for non-subs) and the indices of the m_subs array for subs.